File: aais-middleware/src/aais_tasks/crm_integration.ts
/**
 * Mythic: CRM follow-up from AAIS task
 * Engineering: createFollowUpFromTask
 */
import type { AaisTask } from "./aais_task_model.js";
import type { CrmAdapter } from "../provider_adapters/crm_adapter.js";

export interface CrmFollowUpResult {
  ok: boolean;
  reasonCode: string;
  skipped?: boolean;
  skipReason?: string;
  dealId?: string;
  noteId?: string;
  error?: string;
}

export async function createFollowUpFromTask(
  crm: CrmAdapter,
  task: AaisTask,
  leadId?: string,
): Promise<CrmFollowUpResult> {
  const tags = (task.tags || []).map((t) => t.toLowerCase());
  const wantsCrm = Boolean(leadId) || tags.includes("crm");
  if (!wantsCrm) {
    return {
      ok: true,
      skipped: true,
      skipReason: "no crm leadId/tag",
      reasonCode: "CRM_FOLLOWUP_SKIPPED",
    };
  }
  if (!crm.isConnected()) {
    return {
      ok: false,
      skipped: true,
      skipReason: "CRM not connected",
      reasonCode: "CRM_NOT_CONNECTED",
      error: "CRM store unavailable or disconnected",
    };
  }
  const follow = await crm.createFollowUp(task, leadId || task.id);
  return {
    ok: Boolean(follow.ok),
    reasonCode: String(follow.reasonCode || "CRM_FOLLOWUP"),
    dealId: follow.dealId != null ? String(follow.dealId) : undefined,
    noteId: follow.noteId != null ? String(follow.noteId) : undefined,
    error: follow.error != null ? String(follow.error) : undefined,
  };
}

File: src/aais_tasks/crm_integration.py
"""CRM follow-up from AAIS task.

# Mythic: CRM follow-up from AAIS task
# Engineering: create_follow_up_from_task
"""

from __future__ import annotations

from typing import Any

from src.aais_tasks.aais_task_model import AaisTask
from src.operator_middleware_plugs.adapters.crm_adapter import CrmAdapter


def create_follow_up_from_task(
    crm: CrmAdapter,
    task: AaisTask,
    lead_id: str | None = None,
) -> dict[str, Any]:
    tags = [t.lower() for t in (task.tags or [])]
    wants = bool(lead_id) or "crm" in tags
    if not wants:
        return {
            "ok": True,
            "skipped": True,
            "skip_reason": "no crm leadId/tag",
            "reason_code": "CRM_FOLLOWUP_SKIPPED",
        }
    return crm.create_follow_up(task.to_dict(), lead_id)
